File: src/lib/api/cache-purge.js
/**
 * @flow
 * @format
 */

/**
 * External dependencies
 */
import gql from 'graphql-tag';

/**
 * Internal dependencies
 */
import API from 'lib/api';

const mutation = gql`
	mutation PurgePageCacheMutation(
		$appId: Int!
		$envId: Int!
		$urls: [String!]!
	) {
		purgePageCache( input: {
			appId: $appId
			environmentId: $envId
			urls: $urls
		} ) {
			success
			urls
		}
	}
`;

// The subquery for environments lets users choose any environment, including production.
export const appQuery = `
	id
	name
	environments {
		id
		appId
		name
		primaryDomain {
			name
		}
		type
	}
`;

export async function purgeCache( appId, envId, urls ) {
	const api = await API();

	const variables = {
		appId,
		envId,
		urls,
	};

	const response = await api.mutate( { mutation, variables } );

	const { data: { purgePageCache } } = response;

	return purgePageCache;
}

File: internal/cachepurge/cachepurge.go
// Package cachepurge wraps the PurgePageCache mutation.
//
// Node parity: src/lib/api/cache-purge.ts. The server canonicalizes the
// supplied URLs (e.g. host-normalization) and returns the canonical list on
// the response payload, so callers should use the returned slice for any
// downstream "Purged URL: ..." output rather than echoing the input.
package cachepurge

import (
	"context"

	"github.com/Khan/genqlient/graphql"

	"github.com/Automattic/vip/internal/gql"
)

// Purge invokes the purgePageCache mutation against the given environment
// and returns the server-canonicalized URL list. The returned slice MAY
// differ from urls (server normalizes hosts/casing); callers should rely
// on it when echoing results to the user.
func Purge(ctx context.Context, c graphql.Client, appID, envID int64, urls []string) ([]string, error) {
	input := &gql.PurgePageCacheInput{
		AppId:         appID,
		EnvironmentId: envID,
		Urls:          urls,
	}
	resp, err := gql.PurgePageCache(ctx, c, input)
	if err != nil {
		return nil, err
	}
	if resp == nil || resp.PurgePageCache == nil {
		// Defensive: schema marks PurgePageCachePayload non-null, but a
		// pathological server response could omit it. Return empty so the
		// caller prints nothing rather than panicking.
		return nil, nil
	}
	return resp.PurgePageCache.Urls, nil
}
